src/analysis/processing/qgsalgorithmfiledownloader.cpp

```diff
@@ -0,0 +1,105 @@
+ ***************************************************************************/
+
+#include "qgsalgorithmfiledownloader.h"
+#include "qgsfiledownloader.h"
+#include <QEventLoop>
+#include <QFileInfo>
+#include <QUrl>
+
+///@cond PRIVATE
+
+QString QgsFileDownloaderAlgorithm::name() const
+{
+  return QStringLiteral( "filedownloader" );
+}
+
+QString QgsFileDownloaderAlgorithm::displayName() const
+{
+  return tr( "File downloader" );
+}
+
+QStringList QgsFileDownloaderAlgorithm::tags() const
+{
+  return tr( "file,downloader,internet,url" ).split( ',' );
+}
+
+QString QgsFileDownloaderAlgorithm::group() const
+{
+  return tr( "File tool" );
+}
+
+QString QgsFileDownloaderAlgorithm::shortHelpString() const
+{
+  return tr( "This algorithm downloads a URL on the file system." );
+}
+
+QgsFileDownloaderAlgorithm *QgsFileDownloaderAlgorithm::createInstance() const
+{
+  return new QgsFileDownloaderAlgorithm();
+}
+
+void QgsFileDownloaderAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterString( QStringLiteral( "URL" ), tr( "URL" ), QVariant(), false, false ) );
+  addParameter( new QgsProcessingParameterFileDestination( QStringLiteral( "OUTPUT" ),
+                tr( "File destination" ), QObject::tr( "*.*" ), QVariant(), true ) );
+  addOutput( new QgsProcessingOutputFile( QStringLiteral( "OUTPUT" ), tr( "File destination" ) ) );
+}
+
+QVariantMap QgsFileDownloaderAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  mFeedback = feedback;
+  QString url = parameterAsString( parameters, QStringLiteral( "URL" ), context );
+  QString outputFile = parameterAsFileOutput( parameters, QStringLiteral( "OUTPUT" ), context );
+
+  QEventLoop loop;
+  QgsFileDownloader *downloader = new QgsFileDownloader( QUrl( url ), outputFile, QString(), true );
+  connect( mFeedback, &QgsFeedback::canceled, downloader, &QgsFileDownloader::cancelDownload );
+  connect( downloader, &QgsFileDownloader::downloadError, this, &QgsFileDownloaderAlgorithm::reportErrors );
+  connect( downloader, &QgsFileDownloader::downloadProgress, this, &QgsFileDownloaderAlgorithm::sendProgressFeedback );
+  connect( downloader, &QgsFileDownloader::downloadExited, &loop, &QEventLoop::quit );
+  downloader->startDownload();
+  loop.exec();
+
+  if ( ! QFileInfo( outputFile ).exists() )
+    throw QgsProcessingException( tr( "Output file doesn't exist." ) );
+
+  QVariantMap outputs;
+  outputs.insert( QStringLiteral( "OUTPUT" ), outputFile );
+  return outputs;
+}
+
+void QgsFileDownloaderAlgorithm::reportErrors( QStringList errors )
+{
+  throw QgsProcessingException( errors.join( '\n' ) );
+}
+
+void QgsFileDownloaderAlgorithm::sendProgressFeedback( qint64 bytesReceived, qint64 bytesTotal )
+{
+  if ( bytesTotal != 0 )
+  {
+    int percent = int( bytesReceived / bytesTotal * 100 );
+    mFeedback->setProgress( percent );
+    if ( percent % 10 )
```

Comment:
This value might be too small for small files and too high for big files.
I would either report every change or - if this is too spamy - throttle with a timer.
It would also be very nice to format this to [KB/MB/GB](https://stackoverflow.com/a/30958189/2319028) (or put the percentage before the byte count)

Reply:
I personally think it's ok like this. You still get the finer progress reports, so this is really just a logging/diagnostic helper.